Expose Llava as a shared library for downstream projects

damian0815 opened the pull request:
* [x] convert llava example into struct and init/process/free prompts
* [ ] populate llava-test.cpp with tests
* [x] load image from base64 string embedded in the prompt
* [x] decouple image loading from model loading so a single `llava_context` instance can serve multiple prompt requests
* [ ] whatever else is necessary to enable llama-cpp-python bindings ..?

damian0815 committed:
wip llava python bindings compatibility

damian0815 committed:
add external llava API

FSSRepo commented:
This PR will be very helpful for implementing `llava` on the server.

damian0815 committed:
add base64 in-prompt image support

damian0815 commented:
base64 support is working
```
../llama.cpp/build/bin/llava -m ./ggml-model-q5_k.gguf --mmproj mmproj-model-f16.gguf --temp 0.1 -p '<img src="data:image/jpeg;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAAEACAMAAAA+zbsKAAADAFBMVEX////7+/z/09b/c3stLS3+/v///f2BgYEGBgZaWlr////09PT+/f74+Pj6+vvvikfy8vL9/Pz/+vf5+vxQa5/wjUvyl1v5+fn09fn+/v7/+PSGhob39/hBX5g/Xpf//v7959r29vY9PT3vhT/wiETvhkGhoaHl5eVnZ2f8/P2cnJz//Pr6z7V0irPd3d2Ojo785NbW1tbwj1DZ2dlIZJtKZ51tbW3v8Pb83873+PtYcaTn5+dFYpqJiYn1sYW4uLims82Fl7vxlVlqga7s7/XN1OPp7PPh5e5vhbBDYJljY2NwcHD+9/L+8Oihr8vxkVNbdKXm6fHj5/Bqamr/2tz/9fb//f5ERETymmBTbaF2dnZgYGDh4eH/3d///Pu+vr71r4I9XJb4wJ3znWb+8uv+9e/x8/fe4u3xk1bX3Oj84tL/4OLNzc3607rj4+P96+GqttCurq7U1NS6urqWpcSnp6eUlJSRkZG0v9V9fX3r6+vFxcVISEjKyspOTk75w6L1rX796t6EhIT5yKnzoWz3uZJ4jLX2s4n728eJmr33to6Nnb+YmJj0o3DHx8fV2ueQn8FNaJ6xvdTq6urHz+Bzc3P/+vr/4uWqqqo4WZT+8+372sX1qHjyn2n6zbKsrKykpKSAk7pheqmSosL73crb3+qZqMbEzd7Bytx7e3tdXV3/19r/8PL/5ui2trbS0tL3u5X0pnS+x9oAAABWVlbR1uX/6+397eT3vZn97uVTU1P71r6tudFmfau8vLz/f4X/8/T/1tjQ0NDw8PD/9/h9kLj5x6fT2OYYGBjK0uK6xNjs7OxRUVGzs7P/panCwsLt7e3ugzr72MG4wtedq8hfd6ju7u46Ojr/mZ//eH//wcP/kpn/7u//nqP/6er97ePg4uV5eXmwsLD/0dT/jJP/iI/+yMz1q3v5yqwmJiaZmZn/q7D/ur7/hYz/sretra38/f1AQED/ztEUCwupM0Xyd0L+taf+WlH5TjWOa20MBgbagJHSqbvZV2tsVoWPj7Kei/b2AAAUGUlEQVR42u2daVhT17rH/zmnSHmADJhANlNQwiiThCnM8zxjL2NBQAUpowqoiCgqziNoUdGiOCBtcaR1wF4LdajDccbbc3vbx55zr7bn9ja959x5/BBE1IAJ2SEkWf9Pm+zn2Sv5sda737XWu94XICIiIiIiIiKiScwCBoEgrwYz2QSC3LBmElgKwpIME7OkAAAMMwJmLFgznG+fyxgAQnINcmuYED09Z7DMkKCRDetuTx6zNl5CrbxLLXIfQt0yzmCzG0EjG1ZzMwV2ZiUc6++ZDXLQmZvHGCQDcQxYGR8B1MpGtJ64uLvXEsbtF+Prhwga2bCe2gPLn+QB1FBIzwAAy0U+zgSNbFi9GbfMQs4xBpsdmAEZBczaxWbG9jkEzZtiGxhi0GJZfUYDYO+zRV8/AOUZdfWOAwTNm+Kc5ADsxhA3CWB2cvEAG6AqC+46EDJERERERERERERERES6qW2LCAN5xV7hM4NQkFM1N37cRijIp4DbZ88mkK4ll6j2P589e4Os+ckHa8D+xhP7PAJCPlrUOwbU5De75xPNxPWOgRoaff9vCCy59ZvfElhya87fE1hy670PtATWfI9ZKm/0r/ZoCawm3iWVN/rh77QE1uqs8H3EwL8NFsMJACB44DnXXMWNfvnXGg6LX/1Q6p4arbFrMlFto5/8RsNhLTjkHSW9MungHSGTm3Fh6aVfTr0u7VtVxaXzVNrop/+k8QZ+IfdymtRs7efNnU8M/LiwjJYGHmICAGZds12TT2CN6zo4Vcd18QEA+atY+1XY6Lf/oAV+lnhHXLXTsLdleoEY9PGd0kjXwCi9YbMVnEh61vgefEwEd/iV+NyuSGUzOG1ZotnOjYgBAJivYm0gsMaHRZ0JbBFLzVax6ffEzxp/Is3ot14ifSX6eXbMU++XY1e+vL7VOgVhgd9l1c8AAMEauyKOWueGDc3AL4ul17+MOv/BZgPb7KcCLIhdvc7oAUDierv9AnWuOvRmABnN0mtni5efn1gMOF+cErCQHrFzIQDgwtpwGzUZeGPJMCwOBwCMh2ExGQDgcweQMACA+aLrG1NqgkX1ZcdKX4llrGeqWAkcc6W0swFoDaIQsCw+9y6F3gzgzl3AQf9ckIUF0Fof31PDHLL4MdMR5bWAWW1u/B0GLC0qgy42L1cPLFBRca7RACBwsdukgkmiYKxn1uYw8UMOGLvbhxrjRejNACycYbi7OeBOZjMkde63Fl10kwzG195DzW7gI8e8PMcaiv3kRIiop51SDyxQ1VabmQBgvs62gn6zdeX9MW6IcvPM9GtgvI05eDK+4QWsvMzZoJotQA3cYucZfCEdhjW7IVkZAtzJYLCfNEJSZ8FQEyzwd1j1GwFA0oFwD9obnfPzGDeGEu467J4NbIuPT1hx9wWsxotsqYFf5BOf86R3BNa9JwHAwG0z9sxBwF6fqS5YEIcFLnUCgCReMe2+6ZgGnnJ2bzhhiIbMk9Tg7hFYA+eksIZ+vcNEwktYgzOXAwW3mWqHhegIbh8AoIIVTPdO4gdjxjqUZzjbA50+gGhFwQtYDrdPwjinGW6Zg3BY+RGQUAPU7AbVEwLJTzkc9cOirqceTgcAk/PC5zT7pmNH0VjmxouARSt6e3cbhCDEBwhyhrF7fG/Q7SAYrnD+ovm2PbAlodz4hxzgrsGd9p5tYP86CHQ2qxEWEMWNTQMA87m8S9NobfSr78a8NdBLAWhs/qmyPA+VDcCACDBe3Lx4YAAQBdWJZs8GhuyXsd22AdQi93o3gHmHCYjKJeqERe2yCmMCwAVT0ypaG33/b7VkIj3ayD+yTmYAEGzwPE3rSqBWhhzxN8ftmg6Ac5C1js5d6vf/qIWwEBnB/RoATNYJXWg0W199qo2wUHJ4ZwwA7DvNKqOv0fwrWgmLisqOTQcAm6ws+oz8B1oaU+rUvTdsAQDBBuFc2hYgtDamlKou3CEGIDhot4mu7Z6Pf9ZSWOB3WfczAJivs3OhaQFizFWHSdA8D1XCQqSrd7cRgHlbU+gKRZqmJlKC+U1ZBxJVCQu+l6WhW/tMw+l5JaopppRjU8RinX6Qr1JYetdTD8cAwBHPtbR48uox8ElN4cK5G5RxruWKg5e0ebdEA8g/yFpPB60vv5n8AWj+3FTY4afct5fz0MBDq0N8AEbX7E7REG/6yR8mm9X8imK700pHUckJS3zMOtkJgEmw8KDmvfKNyjqEWRXKu4nyHkdZsNGrnwEgqYO3Qel32Xf/OJmo8qvW80rXJNHg9sh9dmeBq/cZCkDVgVI/ZdudVAOftCbFcxM9kXnyH3QqOcwtAYAyz6wkzYFF7T9g2+FHU1y//LCo66mHSyhAsN9zrZLRNZ/9fpJQzSpbJSyuoG0pTpEjdG1ertEA8l1Yc/dBAyS40MRLKaIxbEoRWFS1dRgFYFaT7TqlNvW//edJgVVRfH9uGZ2xngodzmSGWh8TAzApsi1SxgxMhs0y8euwXbuB3tADxU6y8nfE9U8HMH+TcI3JlIaVdIqXco3uwEUFj/36xnq16QFI7GApsV7z6Z9U3a0qsoTBHrSHLSp6Rjo91vs6BWBeh+2DfExNmfidtl3rp4IHKwqL2u5/+SgAVG31fDBRV/7KxyodgZt4pc9VEjqs8Ol7KsorNhoAkkxTJmo/P1bhSdb5B8NZ66tUEziseKoCo5uBV30BwKaYd2RifUt1Bt7kyAG7rX5GKnr6BPI6iEOtlvABCMpMwyc2Tfzg71T0a6o28Uqfq85hnkgSjOlLrJKnAxDYlKZMiNa0r1TisM9bw+Nt2qfC9f0JZQzhH7JeKgEgKMsqLZsALZVkDDGvKLZbVcYBphgspMVy2ygAnLKUrAnQUkEuGoHNXKGpSyIw9WAh8nD21wAAm3BPxUci7QY+v+qU0LMpUYApCQvb/f3TMTwSFX4n0h3MlnjNlLXJRtWoJg7L6bj38LkCj5TSIwp+zz3f0tqtjmy1O7DBHJiysIA2KyktVBXbXVJsIWTPZzR6Vh6nbbMuzQKmNCyExklDTgU2W3kue9Q0DFev46UUfT5Jk04lYPGPxT0SA4AgaS1LoTwQdBl4wb41pazgqknLxqVMZjZml7WUFuYF3z+vgONMj+sgmHcwXPjMYxKDTJRKY7fgauBS6dW+U7bB8tN670M67HrFabu1FeaAhsBCmqv3cWnWDJMi1tYL8r4U899T3l8vO80ydUkUQHNgIS2C2yad4+9x8cyS1+FSeiJtUrZKaHpt0reYlM0mudDfv0R6xdmfkiJnukBlDfzqTZ6s8zYCaBosKir1cQk1PD3byiqS67+t1OKfic06luf6JHWsaSudp5Rq2zl8oBqYFyx8Js/hxCtzJtwcx+YUj7fJgwNoIixQS61cxcPX85uEB8pUOTz2nSq1DfYzATQUFpyWBl6NfOH7+Jmyit66UPLhxLbC9nicEvJWrRYAmgsL4mrrHQtGjO8q2/Vv+z0TMvD5Hqc8WerrVXTBgrjL+tjIJsH8NbambznM+RvFY0o53wd72q4vMwc0HRb4XYWh4pG/yrbeH79zfaJoYMj8Dc9sSzdVQc2iKWu370ar/ukjfyUVCbNc6JveJro8E5aeslH/BjhdKc59YwNvvtyu27OhmLVq7F+nSDDbrKSDWayUa59PhUTftOWDT3MN7B6Vs8O8KYXV9L1AWQOf6HeeJ9yq6o2ISYeFksvZbaPSwUwrC7Ytfr5HKViJlzp4wuAH8wFtg4WYVG7f6OQ5JvsPsA7I/KXyhHYLElefL2WFF1WZAFoIS2/7Y//rr6QaSjxYbLf10r4JeJGzVh/ssPN8djARU0l01rAw6uNKw5Fe6vs1pcK1Lq93jrcdRzHZP9fUNryobGqhorngB9XHTY15fTQ9L2alFNmYC+R0Ss0vHAn2ZIUH+5lMA7QZFiTHuRElb3iUFc9YvLkHR70axzTwJjYuwab3TYNdkgSYgqK5lAyj23p4N/EVXGXreLbhcy+9GI6yD2dyyprWptiy5lZMJZuuSlhAt9flSFkrBn7B4cL7a5v8Vu97w1edlvi5x/7za21tPQ+cV1kg2pSENX1X4NU0WTemXXhwai3rfvjpTddcfv/9PhMjI47J/CSbIxXX1q3K4t333LrOxcMEU1r0V3RiLi08FCn71p55qy+tO8AS/su/l4abmpqamoaXpvBYQqFp8Bq/qn1Tv3KRKspf9ce1RI5ze0/Vv/7b+fWrOjo6ngWva3LZ8PksaIhUAUscarVxPFqqiynVQFgQh1qHjUdLSxP3TFBOyVZhvmPf/upDAmuU+I+sdoxNa84fCaxXPIhk67H7Fqmc+brdemS9eSxa2pggUdmRWPhyg+w1m/UdgfWajJKtrsruW/QG4GoFLIiTrWX7W6Q0sqyRGFoo098iBl6mqq1cZdAipZFlj8Rqq0NvrkFoTWlkumkttXZ9g9a0PQSWTFFLvS/H6L02kf4dgSVbzJtesUf1iIGXc1bd5p268JX9RI0vjaxCGbX5+0c5jfrgky8JrLH7VlS2fxsFjdekwAJ19HBg98v4La0p2aca6R2N9dolJgZeTqVFxCUTWPLKd+PeruFFCC0pYKtSWscKw6KJgZd36nPMKjZSD8BnpGe9XYyH3Md9EmKz5HS4urn+bQSWvM581OGdD8XEz5JT6S1xSxYQAy+nosOs/48YeHnFT/7f/1lIEVhyzhT/+y/ZbU4Elnz68L9SubvGrN0ocXit25nV3GHoLixBvq+r9ZghXHkrzV79YEazu5nuwrryPnyXxLUslB1sm7dCioYDQGIsAcCRFsW0pHQR1pyfAf7DOP82PRk3Cy7euF2Lgo9ClhmjIf52fB4kP4RQlsvymlckOOggLKkHvzCi8FHkm7gkbpmGEtjnWuRRoh4HSUiPGcd9i8Qwc3fBPcctEt2DNVyyL32jdUvJGMPQPjMPWC5itjauaJXCagDV+dRY92DlD5dG5od6px6fLhvWUwBs53M+T1c4SGG1Aj/kWOoerJH4LL3rEVY7omXCagY49Y73MLhSx2HNeRn5l77R2vX6q29Fh0y2FJZxTifQ+2uebsMavUQjPsPlhvJH3zW+aD8bnUEA1d7T+5N+rohT5ywxXNkK9C7TQZv1amnkkquFV7ePmv1QJ+vuwO0kALPF7r2DjUPUbBHF+MISqCzn6KDNerU0su+uuOzqqTGhmYpvwyuv/m20sCXO9fp0AmscP2v0sk01d2dyJIH1FgM/LKe+lsLYKCMC6w3JLI3sdCa1cGMJg8B6c9VBhqiYJV7+yb4E1qsSjJGFQBwVax1xXExgjdbYpZEZ3YcLY6P4BNa4Bn5EaaH+XmF9BNaIxi2NzEjr2rlzcwmTwJLqLblo9GJ2cK0294kJLLnE6OsK5Ib1SQgsufJnMUs2Z3u5ti0w0nlYckXR6KUfe7w3tr+EIrDkkVPMrseFqUuOGuk0rG/lzSZJ8aOucq0jHsaIdReWIqIWJkcUZm8+E62zPUuhDLiStOMbra1Su64zdBKW4mGS/DNXs/d6be4u4VMElhz96+iusOy9/odC+/i6BWtipZEpfszxzVyrnY/DzkSLGZSuwFLGtz8a2vI4bq//xv6vS3yNdAGWcqWRp6cfDw3z32uV6rq5ui2Noe2wlD7JasSPLjmz47K3t/dOf9djS/uiIxeIpxtpJyy6Dg3wtz9ccjXC33vvXm7Eoa7Qh91fby9J8+UztAoWnWnspkcejequ7mo5vHPv3kKv7NTDEbEtVzd2PQrt33Wz+/jXfdsXlsSkp0VH+vL5YibDSPNg0V4a2Wi6mL8gsiTqZuiSsJaI1OzsbC6Xu3Ont7e3t5eXl5eXV2BgYGBgYFxc3GaNg6XqjCGMBdExR7dHtR0/c3Pprv7q0OTkR8eOLenq2rFj80ONg6XOXDSDDhoGS52nwmb/mNA4Q/atVv236MY7+pOv//iLvtrk8+7ZG08bZW6HnHz3LTp79t3J13/++V316ezZszfqDDVnGNJZGlnhYXjjx4RKTbJZ6swmmRdUbkYMvLxuhRLHgNTjOpAMuAo4pSS3svyioTSy7sAi+eA1xMBrHKyPSRo7+aVEaWTdg6WpUk9M6Z9oeczAuZMwbLgn+6ak3A1grrxDDPzwzKWuEpWOs2XfZC5rBzjuA5oP6xt6nmNISUTx5ZaAxDBghgRgMJlsUMZDy40B9tNOQwksOYCE3WpIAQwmY+gWE6CMb7EZGgRL4dLIsjW4gl3Z8+PKIEhqe3INehlwDnJfhkU+uefiW9H8zswEESPhLsx+Mcg1CKFQk1B/7kkQMODoaGAxpHMGfnAmG3mOJyVwuy3i9J4LwJbbnUMM+184xjm1sDzRzqGYjgUYMFgE0W03tM/sZW/LNIR7EKP1JzfNgaVIaeS3wKp0nA3YJzgEuF28iy3nDEHdCxA1JLhLbRbTsYDq9DEEcmrRvtsSZjMdUNNTfm+QoUE267f0wtLPTEhI8GnEFnfAuDY+p84xaAQWx72eA+jXof0EA5yZlZjR7uhTv1x3YdW5A2BQ2OIOiAwKKDi/hEU5NzOB3NoRWIDZIsd2zYElT2lkOWEFON41o/IuimbM9smTwsodMBStXAZOUL0lh+lYgPJ4kaFoRd4ILOeP2EP6dTpn4Nm5hjD+Kad2hqTzaf3TTjO02wMz6hPqLeyDgAafLSKGfiPManOCfHqBkDoOJLmtGPBxD3oq0iDX4Q+0PIbjJgEGywcYMBNtczMGbrUCGCxvbGVXAmZu225RlWyAM7tBJAEGAyhQbkxwHBpPDlEaZLO+0cyuTNazpvxEmhT8IEs0KhFZ/FNAZFmZGHjViGyFKSCyyaqAyPa9AtLUwJAhthoapWnxj4iIiIiIiIiIiIiIiIiIiIiISAf1/77CZ10v9vU6AAAAAElFTkSuQmCC"> describe the visual content of this image' --verbose-prompt
```
output:
```The image shows a graph with a line that starts at the top left corner and goes down to the bottom right corner. The line is labeled "loss" and has a negative slope. The graph also has a line labeled "validation" that starts at the top right corner and goes down to the bottom left corner. This line has a positive slope. The graph is likely showing the results of a machine learning model's performance, with the "loss" line representing the model's error and the "validation" line representing the model's accuracy.```

monatis commented on examples/llava/base64.hpp:
Probably this should go to `common`.

damian0815 replied:
done

monatis commented on examples/llava/clip.cpp:
```diff
@@ -682,25 +682,39 @@ clip_image_u8 * make_clip_image_u8() { return new clip_image_u8(); }
     memcpy(img->data, data, img->size);
+}
 
+bool clip_image_load_from_bytes(const unsigned char * bytes, int bytes_length, clip_image_u8 * img) {
```
```suggestion
bool clip_image_load_from_bytes(const unsigned char * bytes, size_t size, clip_image_u8 * img) {
```

damian0815 replied:
done

monatis commented on examples/llava/llava.cpp:
```diff
@@ -2,101 +2,166 @@
 
-    const char * clip_path = params.mmproj.c_str();
-    const char * img_path = params.image.c_str();
+static void find_image_tag_in_prompt(const std::string& prompt, size_t& begin_out, size_t& end_out) {
```
Better to have such functions in `llava-utils.h`

damian0815 replied:
👍 done

monatis commented on examples/llava/llava.cpp:
```diff
@@ -2,101 +2,166 @@
+    auto base64_bytes_start = img_base64_str_start + strlen(IMG_BASE64_TAG_BEGIN);
+    auto base64_bytes_count = img_base64_str_end - base64_bytes_start;
+    auto base64_str = prompt.substr(base64_bytes_start, base64_bytes_count );
+    printf("base64_str: '%s'\n", base64_str.c_str());
```
Let's not print this.

damian0815 replied:
was debugging/wip - done.

monatis commented on examples/llava/llava.cpp:
```diff
@@ -2,101 +2,166 @@
 
-    if (!clip_image_preprocess(ctx_clip, &img, &img_res, /*pad2square =*/ true)) {
-        fprintf(stderr, "%s: unable to preprocess %s\n", __func__, img_path);
+struct llava_context * llava_init(gpt_params * params) {
```
Image loading and inference parts should be stripped off this function.

damian0815 replied:
done

monatis commented on examples/llava/llava.cpp:
```diff
@@ -109,28 +174,49 @@ int main(int argc, char ** argv) {
+    free(ctx_llava->image_embd);
+}
+
+void llava_process_prompt(struct llava_context * ctx_llava, gpt_params * params, const char * prompt) {
```
Let's avoid such god-like functions for now --it kills hackability in early stages of development --it brings no benefit over functions in `llava-utils.h`. Better to have single-responsibility functions as much as possible to enhance the development speed and flexibility.

damian0815 replied:
👍 sure

FSSRepo commented:
Server now support LlaVA #3589

damian0815 committed:
wip refactor image loading

damian0815 committed:
refactor image load out of llava init

damian0815 committed:
cleanup

damian0815 committed:
further cleanup; move llava-cli into its own file and rename

damian0815 committed:
move base64.hpp into common/

damian0815 commented:
@monatis thanks for the review, but do please note this PR is still in "draft" state. i addressed all your comments, and actually many of them anticipated what i had planned to do anyway.

so. i'm not sure if the following is correct/desired or in fitting with the way this project is organised, but i went ahead and did it anyway. 
* i have refactored the llava main executable into its own cpp file `llava-cli.cpp`. building now emits `llava-cli` executable rather than `llava`. documentation has been updated to reflect this.
* this leaves `llava.h` and `llava.cpp` as a standalone library. i've updated the CMakeLists.txt to add a `libllava.a` output. again, not sure if this is correct.

as mentioned in the top post my primary goal here is to enable lmql support, which means enabling llava code paths with llama-cpp-python bindings. am i headed in the right direction to do this, or is there something more appropriate i could be doing? i note there's a `server` project as well, but i'm not sure if/how this plays with llama-cpp-python.

damian0815 committed:
collapse clip and llava libraries

damian0815 commented:
actually now that i look over llama-cpp-python this might all be overengineered. all that's really needed i guess is
1. a mechanism to load the multimodal projector model
2. some way of generating image embeds and injecting them into the llama context

damian0815 committed:
move llava into its own subdir

damian0815 committed:
wip

damian0815 committed:
fix bug where base64 string was not removed from the prompt

damian0815 committed:
get libllava to output in the right place

damian0815 committed:
expose llava methods in libllama.dylib

damian0815 committed:
cleanup memory usage around clip_image_*

damian0815 committed:
cleanup and refactor *again*

damian0815 commented:
this is in a much better state now.

the public api is basically:
```
/** load mmproj model */
LLAMA_API struct clip_ctx * clip_model_load(const char * fname, const int verbosity);
/** free mmproj model */
LLAMA_API void clip_free(struct clip_ctx * ctx);

/** sanity check for clip <-> llava embed size match */
LLAMA_API bool llava_validate_embed_size(const llama_context * ctx_llama, const clip_ctx * ctx_clip);

/** build an image embed from image file bytes */
LLAMA_API struct llava_image_embed * llava_image_embed_make_with_bytes(struct clip_ctx * ctx_clip, int n_threads, const unsigned char * image_bytes, int image_bytes_length);
/** build an image embed from a path to an image filename */
LLAMA_API struct llava_image_embed * llava_image_embed_make_with_filename(struct clip_ctx * ctx_clip, int n_threads, const char * image_path);
/** free an embedding made with llava_image_embed_make_* */
LLAMA_API void llava_image_embed_free(struct llava_image_embed * embed);

/** write the image represented by embed into the llama context with batch size n_batch, 
 * starting at context pos n_past. on completion, n_past points to the next position in the context after the image embed. */
LLAMA_API bool llava_eval_image_embed(struct llama_context * ctx_llama, const struct llava_image_embed * embed, int n_batch, int * n_past);
```

damian0815 committed:
update headerdoc

slaren commented:
Some minor suggestions about the API:
- Move `n_threads` to the context as llama does
- Do not ask for `n_batch`, use the value from `llama_context::cparams`. You can add a function `uint32_t llama_n_batch(llama_context * ctx)` to obtain it if needed.

ggerganov reviewed (commented):
The `llava` API is not a good idea.

The only thing a 3rd party project needs in order to implement LLaVA is:

- CLIP API
- LLaMA API

Everything else are helper functions that do not belong neither to `llama.cpp` nor `clip.cpp`. If we want to reuse LLaVA helpers across `llama.cpp`'s examples (e.g. `llava`, `server`, etc.) these should go into `common/llava.h/.cpp`

monatis commented:
> `LLAMA_API struct llava_image_embed * llava_image_embed_make_with_filename(struct clip_ctx * ctx_clip, int n_threads, const char * image_path);`

Some first-impression suggestions from my side:

- Better to keep such structs / functions in `clip_` prefix because:
  1. Having two different prefixes is confusing.
  2. It's not future-proof --I'll implement other SOTA multimodal models very soon starting with Idefics and re-use these functions to encode images.
- Use the verb `encode` instead of `make` --the latter implies a constructor.
- Use `size_t` instead of `int` to type sizes.

I'll make a thorough review tomorrow.

FSSRepo commented:
> The `llava` API is not a good idea.
> 
> The only thing a 3rd party project needs in order to implement LLaVA is:
> 
> * CLIP API
> * LLaMA API
> 
> Everything else are helper functions that do not belong neither to `llama.cpp` nor `clip.cpp`. If we want to reuse LLaVA helpers across `llama.cpp`'s examples (e.g. `llava`, `server`, etc.) these should go into `common/llava.h/.cpp`

See my implementation of llava in server example

Josh-XT commented:
Is this one still being reviewed? llama-cpp-python implementation has a PR waiting for this one.

cebtenzzre commented:
This PR will have to be updated because #3677 was merged without using it.

sorny92 commented on CMakeLists.txt:
```diff
@@ -698,6 +698,7 @@ target_include_directories(llama PUBLIC .)
 target_compile_features(llama PUBLIC cxx_std_11) # don't bump
 target_link_libraries(llama PRIVATE
     ggml
+    llava
```
```suggestion
    -Wl,--whole-archive llava
    -Wl,--no-whole-archive
```

sorny92 commented on CMakeLists.txt:
```diff
@@ -707,6 +708,22 @@ if (BUILD_SHARED_LIBS)
+        LINK_FLAGS "-Wl,-all_load"
+    )
+    else ()
+        set_target_properties(llama PROPERTIES
```
I removed this clausule as the code I added on line 701 is more restrictive

sorny92 commented on examples/llava/CMakeLists.txt:
```diff
@@ -1,19 +1,7 @@
+target_link_libraries(${TARGET} PRIVATE common llama llava ${CMAKE_THREAD_LIBS_INIT})
 target_compile_features(${TARGET} PRIVATE cxx_std_11)
 if(TARGET BUILD_INFO)
     add_dependencies(${TARGET} BUILD_INFO)
```
```suggestion
    add_dependencies(${TARGET} BUILD_INFO)
    if (BUILD_SHARED_LIBS)
    set_target_properties(${TARGET} PROPERTIES POSITION_INDEPENDENT_CODE ON)
endif()

```

sorny92 reviewed (commented):
Hi, great job here! I wanted to test https://github.com/abetlen/llama-cpp-python/pull/821 and had some problems to build it. 
Mostly I got some multiple definition issues at the linking stage.

I add some comments with the changes I had to do to make it run. Now it works in my linux system!

ggerganov commented:
Can we move the files from the `llava` folder into `examples/llava` and make the example produce a `llava` library that exposes the LLaVA API?

I prefer not to merge the LLaVA API into the main `llama` library at this point. If we can isolate everything as a community maintained example, we should be able to merge this change. `llama-cpp-python` for example would have to link both to the `llama` and `llava` libs produced by `llama.cpp`

ggerganov commented:
Bumping this because the [llama-cpp-python](https://github.com/abetlen/llama-cpp-python/issues/813) project will likely be OK to link to a separate `llava.cpp` library that we can build as part of the examples. Long-term, the better solution would be to support CLIP natively as arch from `llama.cpp`, but this can take more time (see @monatis's explanation here: https://github.com/abetlen/llama-cpp-python/issues/813#issuecomment-1789221323)

Not sure if I should make a separate issue to track this, or we just finish the PR here. Seems like this can be achieve relatively easily. Looking for volunteers to help with this

monatis commented:
I think this PR can be updated so as to move the `llava` directory back to `examples` and fix conflicts --hope that they are minor ones. If @damian0815, the original author of this PR, does not have time to do so I can go on with it.

damian0815 commented:
time is a bit tight at the moment, please go ahead @monatis

monatis committed:
Move llava back to examples

monatis committed:
build with cmake, not tested (WIP)

monatis commented:
Started to work on this, will continue tomorrow.

monatis committed:
Editorconfig

monatis committed:
Merge branch 'master' into llava-lib

monatis committed:
Editorconfig

monatis committed:
Build with make

monatis committed:
Build with make

monatis committed:
Fix cyclical depts on Windows

monatis committed:
attempt to fix build on Windows

monatis committed:
Merge branch 'master' into llava-lib

monatis commented:
Anyone knowing how to fix build on Windows? It builds on other platforms but fails on windows with a linkage error that I couldn't relate to this PR.

cebtenzzre commented on CMakeLists.txt:
```diff
@@ -682,6 +682,22 @@ if (BUILD_SHARED_LIBS)
+    # are not used. This changes that.
+    if (WIN32)
+        set_target_properties(llama PROPERTIES
+        LINK_FLAGS "/WHOLEARCHIVE"
```
This flag apparently breaks MSVC unless you also set `/FORCE:MULTIPLE`. I'm not sure why it's spelled "WIN32" here, as I believe MinGW has its own linker.

monatis replied:
Thanks, it works

monatis committed:
attempt to fix build on Windows

monatis committed:
Upd TODOs

monatis committed:
attempt to fix build on Windows+CUDA

monatis reviewed (changes requested):
Windows+CUDA still fails.

monatis committed:
Revert changes in cmake

ggerganov commented on build-info.h:
I don't think `build-info.h` should be source controlled

ggerganov commented on common/CMakeLists.txt:
```diff
@@ -41,6 +41,7 @@ endif()
 set(TARGET common)
 
 add_library(${TARGET} STATIC
+base64.hpp
     common.h
```
```suggestion
add_library(${TARGET} STATIC
    base64.hpp
    common.h
```

ggerganov commented on examples/llava/llava-utils.h:
```diff
@@ -1,14 +1,21 @@
+#ifndef LLA_UTILS_H
```
```suggestion
#ifndef LLAVA_UTILS_H
```

ggerganov commented on examples/llava/llava-utils.h:
```diff
@@ -1,14 +1,21 @@
+#ifndef LLA_UTILS_H
+#define LLAVA_UTILS_H
+
 #pragma once
```
`#pragma once` not needed with `ifndef` guard

ggerganov commented on examples/llava/llava.cpp:
```diff
@@ -2,163 +2,166 @@
 #include "llava-utils.h"
```
The way it is now, `llava-utils.h` is both a private and public header.

Is `llava-utils.h` used in `llava.cpp`? If not, remove the include here.
If it is used, it's better to move whatever functionality is needed from `llava-utils.h` into `llava.cpp`. This way, `llava-utils.h` can remain as something similar to what `common` is to `llama.cpp` and be used only by the examples (i.e. `llava-cli` in this case).

ggerganov commented on examples/llava/llava.h:
```diff
@@ -0,0 +1,36 @@
+/** free an embedding made with llava_image_embed_make_* */
+
+/** write the image represented by embed into the llama context with batch size n_batch, starting at context pos n_past. on completion, n_past points to the next position in the context after the image embed. */
+LLAMA_API bool llava_eval_image_embed(struct llama_context * ctx_llama, const struct llava_image_embed * embed, int n_batch, int * n_past);
```
Ideally, the `llava` lib should not reuse the `LLAMA_API` export macros, but have it's own `LLAVA_API`.
But I guess we can leave it like this for now to make it simpler.

ggerganov commented on examples/llava/llava.h:
```diff
@@ -0,0 +1,36 @@
+#define LLAVA_H
+
+#include "ggml.h"
+#include "common.h"
```
`llava.h` should not include `common.h`

ggerganov commented on examples/llava/clip.h:
```diff
@@ -1,7 +1,7 @@
 #define CLIP_H
 
-#include "ggml.h"
+#include "llama.h"
```
`clip.h` includes `llama.h` which is not great. The only reason is to be able to reuse the `LLAMA_API`.
Can be improved by having dedicated `CLIP_API`.

monatis replied:
Thanks for the review --I handled them and it should be ready as a shared lib now.

ggerganov reviewed (commented):
This is very close to what I had in mind, though some things on the API level can be improved. I've added a few comments regarding that.

monatis committed:
Fix according to review comments

abetlen commented on examples/llava/CMakeLists.txt:
```diff
@@ -1,14 +1,22 @@
-target_link_libraries(${TARGET} PRIVATE common ggml ${CMAKE_THREAD_LIBS_INIT})
+set(TARGET llava)
+
+add_library(${TARGET} STATIC llava.cpp llava.h clip.cpp clip.h)
```
`STATIC` should be dropped here to allow building `llava` as a shared library.

monatis replied:
Yes, will update it to build a shared library as well.

abetlen replied:
Great, from my tests that looks like the only change I need to integrate this with `llama-cpp-python`.

monatis replied:
Good, now it also builds as a shared lib.

abetlen replied:
That works, thank you!

Ommiting the `STATIC` / `OBJECT` from `target_link_libraries` also seemed to work with just the `llava` target without the need for the `llava_shared` target but if having different targets is preffered I don't mind it.

FSSRepo commented:
It seems to me that a good way to check if this PR is achieving its goals is to implement the Llava API in the server example. If it proves to be a better option than how Llava currently functions on the server, it's easy to deduce that we're heading in the right direction since we can eliminate a lot of duplicate code from the server. This is just my opinion to help; I don't intend to come across as a know-it-all.

monatis committed:
Support building as a shared library

cebtenzzre committed:
address review comments

damian0815 commented:
glad this made it in, thanks folks!
